### PR TITLE
[AS-172] Terraform for import service

### PR DESCRIPTION
**Pull request by helgridly**

### CODE REVIEW

File: import-service/app-engine.tf [hunk 1]
```diff
@@ -0,0 +1,26 @@
+
+resource "google_app_engine_application" "gae_import_service" {
+  project     = module.import-service-project.project_name
+  location_id = "us-central"
+}
+
+# firewall rules that allow broad ips to access GAE
+# This doesn't work, see https://github.com/terraform-providers/terraform-provider-google/issues/5681
```
Comment (helgridly) [inline import-service/app-engine.tf]: This is going to be a problem. I can write a script to create the firewall rules, I guess.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,19 @@
...
++ create import service SA
++ give import SA pubsub admin and token creator permissions in terra-importservice-env
++ give terra-importservice-env GAE SA token creator permissions on import service SA
++ database
++ pubsub topic and subscription
++ create bucket for translated imports
++ firewall around GAE
+
+
+manual (unless terraform can run arbitrary sh):
```
Comment (gmalkov) [inline README.md]: It can, an example of a script that was recently incorporated into TF is here:
https://github.com/broadinstitute/terraform-terra/pull/125

Comment (gmalkov) [reply README.md]: The tricky bit is making sure it's idempotent and fails in a recoverable manner.

File: import-service/app-engine.tf (hunk 1 above)
Comment (helgridly) [reply import-service/app-engine.tf]: update: i did

File: README.md [hunk 3]
```diff
@@ -1,2 +1,25 @@
...
+`terra_google_project`                          | google project that terra monolithic services run in (i.e. where rawls and sam SAs live)
+
+
+Your .tfvars MAY contain the following, if you want to override the defaults during local development:
+
+
+**variable**|**default**|**description**
+-----|-----|------
+`import_service_google_project` | `{var.env}` | The Google project to create, `terra-importservice-{import_service_google_project}`
+`bucket_suffix` | `{var.env}` | Creates the import service batchUpsert bucket, `importservice-batchupsert-{bucket_suffix}`
```
Comment (davidangb) [inline README.md]: defaults for `import_service_google_project` and `bucket_suffix` seem wrong

Comment (helgridly) [reply README.md]: the default i listed here is the default value for the variable, not the final value of the generated string. it's easier to see for the bucket suffix (if you do nothing, the bucket will be called `importservice-batchupsert-yourenv`) but i see now that this is all a bit confusing. i'll update some things, thank you

Comment (helgridly) [reply README.md]: wow holy shit this is garbage

Comment (helgridly) [reply README.md]: okay, try again? i think bucket_suffix is fine, tell me why i'm wrong? (never mind, just noticed missing dash)

**Review by gmalkov — APPROVED** [review]

File: atlantis.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+version: 3
+projects:
+  - name: import-service
+    dir: import-service
+    workflow: import-service
+    workspace: gmalkov
```
Comment (davidangb) [inline atlantis.yaml:6]: is `gmalkov` the right value?

File: import-service/google-project.tf [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    google.target = google.target
+    vault = vault
+  }
+}
+
+locals {
+  import_service_sa_email = "import-service@${module.import-service-project.project_name}.iam.gserviceaccount.com"
+}
+
+# Give the GCP pubsub SA permission to send pubsub messages as the import service SA
```
Comment (davidangb) [inline import-service/google-project.tf]: nit: since permissions are so complex/confusing, could you reword this to something like `Give the GCP pubsub SA permission to impersonate the import service SA, which it will use to send pubsub messages as the import service SA"

File: import-service/google-project.tf [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+resource "google_service_account_iam_member" "grant_gcp_compute_sa_roles_on_import_sa" {
+  service_account_id = "projects/${module.import-service-project.project_name}/serviceAccounts/${local.import_service_sa_email}"
+  member = "serviceAccount:service-${module.import-service-project.project_number}@gcp-sa-pubsub.iam.gserviceaccount.com"
+  role = "roles/iam.serviceAccountTokenCreator"
+  depends_on = [module.import-service-project]
+}
+
+# This next section is about granting the Terraform SA permission to impersonate the import service SA. This is needed
+# to reference the import service SA with Sam. There's no way I can find of getting the Terraform SA email address,
+# so we take their token, pass it to Google's tokeninfo, and extract it from there.
```
Comment (davidangb) [inline import-service/google-project.tf:58]: what generates the Terraform access token? Does that process have access to the email address?

File: import-service/google-project.tf [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+  depends_on = [module.import-service-project, google_app_engine_application.gae_import_service]
+}
+
+# Google IAM is a bit "eventually" so add a sleep between granting token creator and retrieving the token.
+resource "null_resource" "delay_before_getting_import_service_token" {
+  depends_on = [google_service_account_iam_member.grant_self_token_creator_on_import_sa]
+  provisioner "local-exec" {
+    command = "sleep 60"
+  }
+}
```
Comment (davidangb) [inline import-service/google-project.tf:93]: ewww. though I don't have a better idea.

File: import-service/output.tf [hunk 8]
```diff
@@ -0,0 +1,8 @@
+output "ATTENTION" {
+  value = <<EOF
+THIS PROFILE REQUIRES MANUAL STEPS!
+To run the manual steps run these scripts in the import-service directory:
+./setup_gae_firewall.py ${module.import-service-project.project_name} ${var.env}
+./add-import-service-sa-to-sam.sh ${var.env} ${data.google_service_account_access_token.sam_access_token.access_token} ${local.import_service_sa_email} ${data.google_service_account_access_token.import_service_token.access_token}
+EOF
+}
```
Comment (davidangb) [inline import-service/output.tf:8]: I feel a bit wonky about outputting access tokens here. Thinking out loud ... is  there any risk to this? Whoever is running these scripts has a lot of privileged access anyway. Are there cases in which this output could be logged?

File: import-service/provider.tf [hunk 9]
```diff
@@ -0,0 +1,7 @@
+provider "google" {
+  alias = "target"
+}
+
+provider "google-beta" {
+  alias = "target"
+}
```
Comment (davidangb) [inline import-service/provider.tf:7]: are these necessary? Are these here to support future changes in which we add more config to these providers?

File: import-service/pubsub.tf [hunk 10]
```diff
@@ -0,0 +1,51 @@
+resource "google_pubsub_topic" "import-service-pubsub-topic" {
+  project = module.import-service-project.project_name
+  name = "import-service-notify"
+}
```
Comment (davidangb) [inline import-service/pubsub.tf:4]: shouldn't this be `import-service-notify-${var.env}`? (or some other syntax flavor of $ENV)

File: import-service/pubsub.tf [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+resource "google_pubsub_subscription" "import-service-pubsub-subscription" {
+  project = module.import-service-project.project_name
+  name  = "import-service-notify-push"
+  topic = google_pubsub_topic.import-service-pubsub-topic.name
+
+  ack_deadline_seconds = 600
+
+  expiration_policy {
+    ttl = ""
+  }
```
Comment (davidangb) [inline import-service/pubsub.tf:17]: same here ... should subscription include ENV?

File: import-service/setup_gae_firewall.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+cmd = f'gcloud --project broad-dsde-{args.env} compute instances list'.split() + [f'--filter=name:gce-firecloud-orchestration-{args.env}* AND status:RUNNING', '--format=json']
+res = subprocess.run(cmd, stderr=subprocess.STDOUT, stdout=subprocess.PIPE, encoding='utf-8')
+check_error(res, "Error finding orch instances, exiting")
+
+orchestrations = json.loads(res.stdout)
+
+orch_inst_ips = get_gae_public_ips(orchestrations)
+for (idx, orch_inst) in enumerate(orch_inst_ips):
+    add_gae_firewall_rule(1000 + ip_count + idx, orch_inst_ips[orch_inst], orch_inst)
+
```
Comment (davidangb) [inline import-service/setup_gae_firewall.py:158]: It would be ideal if this script collected all of the deletes and creates it intends to issue, then only applies those at the end. Else, an error mid-script could result in deleting all firewall rules and re-adding none, leaving things in an ugly state until  everything is debugged and fixed. I'm trying to decide if this is worth holding the PR for ...  what are your thoughts?

**Review by davidangb — COMMENTED** [review]
first-pass comments inline!

File: atlantis.yaml (hunk 4 above)
Comment (helgridly) [reply atlantis.yaml:6]: doesn't super matter -- this is a terraform workspace, i.e. the place where terraform keeps its state. it should probably be updated to something more official-looking once this is going to be run on real environments

File: import-service/google-project.tf (hunk 6 above)
Comment (helgridly) [reply import-service/google-project.tf:58]: The Terraform access token is accessible from the `google_client_config` data block. The insides of that block are opaque.

File: import-service/google-project.tf (hunk 7 above)
Comment (helgridly) [reply import-service/google-project.tf:93]: this has caused me at least a week's worth of anguish

File: import-service/output.tf (hunk 8 above)
Comment (helgridly) [reply import-service/output.tf:8]: the biggest risk is that it gets output into the github PR. however, this repo is private, and the tokens expire after 300s.

File: import-service/provider.tf (hunk 9 above)
Comment (helgridly) [reply import-service/provider.tf:7]: i was repeating incantations i don't understand. @gmalkov  ?

File: import-service/pubsub.tf (hunk 10 above)
Comment (helgridly) [reply import-service/pubsub.tf:4]: no. each import service instance lives in its own google project. the google project has the name of the env in it; there's no need to put the env in the topic name too, because there will only be one topic per project.

File: import-service/pubsub.tf (hunk 11 above)
Comment (helgridly) [reply import-service/pubsub.tf:17]: same response

File: import-service/setup_gae_firewall.py (hunk 12 above)
Comment (helgridly) [reply import-service/setup_gae_firewall.py:158]: it's probably not a hugely intrusive change to the script to do so -- just means generating the list of rawls and orch IPs ahead of the first delete. i'll see what i can do

Comment (helgridly) [reply import-service/setup_gae_firewall.py:158]: done

File: import-service/pubsub.tf (hunk 10 above)
Comment (davidangb) [reply import-service/pubsub.tf:4]: But this isn't what Rawls uses: https://github.com/broadinstitute/firecloud-develop/blob/dev/base-configs/rawls/rawls.conf.ctmpl#L341. You previously said "in the long run import service in dev will be deployed using the terraform i'm writing and i'll make sure that rawls configs are adjusted to point to the right queues as part of that" ... will there be a change to Rawls config, or a change here?

Comment (helgridly) [reply import-service/pubsub.tf:4]: uuuurgh. good catch. i'll make a change to the rawls conf. edit:\ no i won't, i'll do it here

**Review by davidangb — APPROVED** [review]
lgtm; I did not test
